pymc_experimental/utils/model_fgraph.py

```diff
@@ -0,0 +1,174 @@
+    if any(v is not None for v in model.rvs_to_initial_values.values()):
+        raise NotImplementedError("Cannot convert models with non-default initial_values")
+
+    # We start the `dict` with mappings from the value variables to themselves,
+    # to prevent them from being cloned.
+    memo = {v: v for v in values}
+
+    fgraph = FunctionGraph(
+        outputs=rvs + potentials + deterministics,
+        clone=True,
+        memo=memo,
+        copy_orphans=False,
+        copy_inputs=False,
+    )
+    fgraph.coords = coords
+
+    # Introduce dummy `ModelVar` Ops
+    free_rvs_to_values = {memo[k]: v for k, v in rvs_to_values.items() if k in free_rvs}
+    observed_rvs_to_values = {memo[k]: v for k, v in rvs_to_values.items() if k not in free_rvs}
+    potentials = [memo[k] for k in potentials]
+    deterministics = [memo[k] for k in deterministics]
+
+    vars = fgraph.outputs
+    new_vars = []
+    for var in vars:
+        dims = vars_to_dims.get(var.name, None)
+        if var in free_rvs_to_values:
+            new_var = free_rv(var, free_rvs_to_values[var], dims)
+        elif var in observed_rvs_to_values:
+            new_var = observed_rv(var, observed_rvs_to_values[var], dims)
+        elif var in potentials:
+            new_var = potential(var, dims)
+        elif var in deterministics:
+            new_var = deterministic(var, dims)
+        else:
+            raise RuntimeError(f"Variable is not RV, Potential nor Deterministic: {new_var}")
+        new_vars.append(new_var)
+
+    toposort_replace(fgraph, tuple(zip(vars, new_vars)))
+    return fgraph
+
+
+def model_from_fgraph(fgraph: FunctionGraph) -> Model:
+
+    model = Model(coords=getattr(fgraph, "coords", None))
+
+    # Get rid of dummy `ModelVar` Ops
+    fgraph = fgraph.clone()
+    model_vars_to_vars = {
+        model_node.outputs[0]: model_node.inputs[0]
+        for model_node in fgraph.apply_nodes
+        if isinstance(model_node.op, ModelVar)
+    }
+    toposort_replace(fgraph, tuple(model_vars_to_vars.items()))
+
+    # Populate new PyMC model mappings
+    for model_var in model_vars_to_vars.keys():
+        if isinstance(model_var.owner.op, FreeRV):
+            var, value, *dims = model_var.owner.inputs
+            model.free_RVs.append(var)
+            model.create_value_var(var, transform=None, value_var=value)
+            model.set_initval(var, initval=None)
+        elif isinstance(model_var.owner.op, ObservedRV):
+            var, value, *dims = model_var.owner.inputs
+            model.observed_RVs.append(var)
+            model.create_value_var(var, transform=None, value_var=value)
+        elif isinstance(model_var.owner.op, Potential):
+            var, *dims = model_var.owner.inputs
+            model.potentials.append(var)
+        elif isinstance(model_var.owner.op, Deterministic):
+            var, *dims = model_var.owner.inputs
+            model.deterministics.append(var)
+        else:
+            continue  # Raise?
```

Comment:
How would you ever reach this?

Reply:
I was just dumping code. I guess if a user defined a new type of `ModelVar`